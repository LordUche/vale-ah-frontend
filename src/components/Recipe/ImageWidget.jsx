import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * @class ImageWidget
 */
class ImageWidget extends Component {
  state = {
    imageContainerRevealed: false,
  };

  toggleContainerOverlay = () => {
    const { imageContainerRevealed } = this.state;

    this.setState({ imageContainerRevealed: !imageContainerRevealed });
  };

  /**
   * @returns {JSX.Element} Image widget
   */
  render() {
    const {
      props: { images },
      state: { imageContainerRevealed },
    } = this;

    return (
      <div className="image-selector">
        <div className="img-holder">
          <img
            src={
              images.length
                ? images[0]
                : 'https://intl.9point8.ca/image/catalog/no_image_220_220.png'
            }
            alt="placeholder"
            className={`recipe-image${images.length > 1 ? ' more-images' : ''}`}
          />
          {images.length > 1 ? (
            <div
              className="overlay"
              onClick={this.toggleContainerOverlay}
              onKeyDown={e =>
                e.keyCode === 13 ? this.toggleContainerOverlay : null
              }
              role="button"
              tabIndex={0}
            >
              1 + {images.length - 1} more
            </div>
          ) : (
            ''
          )}
          {imageContainerRevealed ? (
            <div className="images-container">
              {images.map(i => (
                <img
                  src={i}
                  alt="recipe"
                  key={Date.now() + Math.random()}
                  className="recipe-images"
                />
              ))}
            </div>
          ) : (
            ''
          )}
        </div>
      </div>
    );
  }
}

ImageWidget.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default ImageWidget;
