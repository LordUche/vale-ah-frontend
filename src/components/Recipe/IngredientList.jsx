import React from 'react';
import PropTypes from 'prop-types';
import { generateKey } from '../../utils/helpers';

/**
 * @returns {JSX.Element} Ingredient list
 */
const IngredientList = ({
  ingredients = [],
  syncRecipeIngredents,
  errorHandler,
}) => {
  /**
   *
   * @param {Event} e
   * @returns {undefined}
   */
  const updateList = e => {
    const {
      target: { value: ingredient },
    } = e;

    if (e.keyCode === 13) {
      e.preventDefault();
      if (e.target.value.length > 0) {
        syncRecipeIngredents([
          ...ingredients,
          { key: generateKey(), ingredient },
        ]);

        e.target.value = '';
      }
    }
  };

  /**
   * Remove ingredient from list
   * @param {number} index index of ingredient to remove
   * @returns {undefined}
   */
  const deleteIngredient = index => {
    const updatedList = [...ingredients];
    updatedList.splice(index, 1);
    return syncRecipeIngredents(updatedList);
  };

  const ingredientList =
    ingredients.length > 0 ? (
      ingredients.map(({ key, ingredient }, index) => (
        <li key={key} className="ingredient-item">
          {index + 1}
          {'.'} {ingredient}
          <i
            className="fas fa-minus remove-icon"
            role="button"
            tabIndex={0}
            onClick={() => deleteIngredient(index)}
            onKeyDown={({ keyCode }) =>
              keyCode === 13 ? deleteIngredient(index) : null
            }
          />
        </li>
      ))
    ) : (
      <div className="flex">
        <img
          src="https://cdn2.iconfinder.com/data/icons/outline-web-application-1/20/cart-512.png"
          className="emp"
          alt="placeHolder"
        />
        <small className="no-ing">No ingredient added yet</small>
      </div>
    );

  return (
    <div className="ingredient-section">
      <p className="col-title">Ingredients</p>
      <div>
        <ul className="ingredient-lists">{ingredientList}</ul>

        <div className="input-block">
          <input
            name="ingredient"
            className="text-field text-field--small ingredient-field"
            placeholder="Add an ingredient (Press Enter)"
            onKeyDown={updateList}
          />
          <small className="err-msg">{errorHandler()}</small>
        </div>
      </div>
    </div>
  );
};

IngredientList.propTypes = {
  syncRecipeIngredents: PropTypes.func.isRequired,
  ingredients: PropTypes.arrayOf(PropTypes.object).isRequired,
  errorHandler: PropTypes.func.isRequired,
};

export default IngredientList;
