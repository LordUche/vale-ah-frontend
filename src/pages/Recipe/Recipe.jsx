import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

/**
 * @description - renders the recipes
 * @param {object} props
 * @returns {jsx} - returns all the recipes if found
 */
const Recipes = ({ recipes }) =>
  Object.values(recipes).map(recipe => (
    <div key={recipe.id}>
      <h4>{recipe.title}</h4>
      <p>{recipe.body}</p>
    </div>
  ));

Recipes.propTypes = {
  recipes: PropTypes.objectOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      body: PropTypes.string,
    })
  ).isRequired,
};

/**
 * @description - maps the state to prop
 * @param {object} state
 * @returns {state} - state
 */
const mapStateToProps = state => ({
  recipes: state.recipes.recipes,
});

export default connect(mapStateToProps)(Recipes);
