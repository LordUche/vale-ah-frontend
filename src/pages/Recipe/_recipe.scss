body {
  background: $color;
  color: $white-color;
}
