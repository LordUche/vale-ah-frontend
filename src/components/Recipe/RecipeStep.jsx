import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { openFilePicker } from '../../utils/helpers';
import ImageWidget from './ImageWidget';

/**
 *
 * @class RecipeStep
 */
class  RecipeStep extends Component {
  state = { addHover: false };

  filePickerRef = React.createRef();

  /**
   * @returns {object} media upload options
   */
  get uploadOptions() {
    const { index } = this.props;

    return {
      stepIndex: index,
      typeFilter: ['image', 'video'],
      uploadType: 'recipe_step',
    };
  }

  /**
   * @param {Event} e
   * @returns {undefined}
   */
  handleDescriptionChange = e => {
    const { syncStepProp, index } = this.props;

    syncStepProp(e.target.value, index, 'description');
  };

  /**
   * Toggle step
   * @returns {undefined}
   */
  toggleAddHover = () => {
    const { addHover } = this.state;

    this.setState({ addHover: !addHover });
  };

  /**
   * append new step Block to DOM
   * @param {Event} e
   * @returns {undefined}
   */
  spawnStep = e => {
    const { addNewStep, description } = this.props;

    if (description.trim() === '' || (e.keyCode && e.keyCode !== 13)) return;

    addNewStep();
  };

  /**
   * open file picker
   * @param {Event} e
   * @returns {undefined}
   */
  openFilePicker = e => {
    openFilePicker(e, this.filePickerRef);
  };

  /**
   * open file picker
   * @param {Event} e
   * @returns {undefined}
   */
  filePickerOnchange = e => {
    const { handleSelectedFiles } = this.props;

    handleSelectedFiles(e, this.uploadOptions);
  };

  /**
   * @returns {JSX.Element} Recipe step component
   */
  render() {
    const {
      props: { description, images, stepNumber, removeStep },
      state: { addHover },
    } = this;

    return (
      <Fragment>
        <div className="ingredient-section">
          <input
            type="file"
            ref={this.filePickerRef}
            onChange={this.filePickerOnchange}
            multiple
            hidden
          />
          <div className="step-description">
            <small className="step-number">Step {stepNumber}</small>
            <textarea
              rows={3}
              className="text-field text-field--large description-field"
              placeholder="Add a new preparation step"
              onChange={this.handleDescriptionChange}
              value={description}
            />
            <div className="step-icon-holder">
              <i
                className="fas fa-image icon-main"
                title="add photos/videos"
                onClick={this.openFilePicker}
                onKeyDown={this.openFilePicker}
                role="button"
                tabIndex={0}
              />
              <i
                className="fas fa-minus icon-main"
                title="add photos/videos"
                onClick={removeStep}
                onKeyDown={removeStep}
                role="button"
                tabIndex={0}
              />
            </div>
          </div>
          <div className="add-step">
            {!addHover ? (
              <p className="add-step-input-line">&nbsp;</p>
            ) : (
              <p className="add-step-input">Add Step</p>
            )}
            <i
              className="fas fa-plus add-icon"
              onMouseEnter={this.toggleAddHover}
              onMouseLeave={this.toggleAddHover}
              onClick={this.spawnStep}
              onKeyDown={this.spawnStep}
              role="button"
              title="Fill the text input before adding a new step"
              tabIndex="0"
            />
          </div>
        </div>
        <ImageWidget images={images} className="image-widget" />
      </Fragment>
    );
  }
}

RecipeStep.propTypes = {
  addNewStep: PropTypes.func.isRequired,
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
  index: PropTypes.number.isRequired,
  stepNumber: PropTypes.number.isRequired,
  syncStepProp: PropTypes.func.isRequired,
  description: PropTypes.string.isRequired,
  removeStep: PropTypes.func.isRequired,
  handleSelectedFiles: PropTypes.func.isRequired,
};

export default RecipeStep;
