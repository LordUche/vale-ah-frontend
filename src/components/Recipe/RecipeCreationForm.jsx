import React, { Component } from 'react';
import PropTypes from 'prop-types';
import RecipeStep from './RecipeStep';
import IngredientList from './IngredientList';
import {
  generateKey,
  openFilePicker,
  fileInputChangeHandler,
} from '../../utils/helpers';

/**
 * Render form to create new recipe
 * @class RecipeCreationForm
 */
class RecipeCreationForm extends Component {
  state = {
    title: '',
    ingredients: [],
    steps: [
      {
        images: [],
        videos: [],
        description: '',
        key: generateKey(),
      },
    ],
    videoList: [],
    cookingTime: '',
    preparationTime: '',
  };

  filePickerRef = React.createRef();

  /**
   * @returns {undefined}
   */
  componentDidMount() {
    const filePicker = this.filePickerRef.current;
    filePicker.addEventListener('change', e =>
      fileInputChangeHandler(e, {
        typeFilter: ['video'],
        uploadType: 'recipe_video',
      })
    );

    window.addEventListener(
      'video_upload_complete',
      ({
        detail: {
          mediaInfo: [{ url }],
        },
      }) => {
        this.setState({ videoList: [url] });
      }
    );

    window.addEventListener('step_upload_complete', ({ detail }) => {
      const { syncStepProp } = this;
      detail.mediaInfo.forEach(({ resType, url }) => {
        if (resType === 'image') syncStepProp(url, detail.stepIndex, 'images');
        if (resType === 'video') syncStepProp(url, detail.stepIndex, 'videos');
      });
    });
  }

  /**
   * @param {number} stepInsertionIndex
   * @returns {undefined}
   */
  addNewStep = stepInsertionIndex => {
    const { steps } = this.state;
    const newSteps = [...steps];

    newSteps.splice(stepInsertionIndex + 1, 0, {
      images: [],
      videos: [],
      description: '',
      key: generateKey(),
    });

    this.setState({ steps: newSteps });
  };

  /**
   * @param {Event} e change event
   * @returns {undefined}
   */
  syncField = e => {
    const { name, value } = e.target;

    this.setState({ [name]: value });
  };

  /**
   * @param {Array} ingredients updated ingredient list
   * @returns {undefined}
   */
  syncRecipeIngredents = ingredients => {
    this.setState({ ingredients });
  };

  /**
   * @param {*} value field value
   * @param {number} index step index
   * @param {string} prop property to update
   * @returns {undefined}
   */
  syncStepProp = (value, index, prop) => {
    const { steps } = this.state;
    const syncedSteps = [...steps];

    if (['images', 'videos'].includes(prop)) {
      syncedSteps[index][prop].push(value);
    } else syncedSteps[index][prop] = value;
    this.setState({ steps: syncedSteps });
  };

  /**
   * @param {Event} e change event
   * @returns {undefined}
   */
  handleSubmit = e => {
    const { handleCreation } = this.props;
    e.preventDefault();
    handleCreation(this.state);
  };

  /**
   * @param {field}  field
   * @returns {undefined}
   */
  handleError = field => {
    const { recipeCreation } = this.props;
    if (recipeCreation.errors[field]) {
      return recipeCreation.errors[field][0];
    }

    return '';
  };

  /**
   * open file picker
   * @param {Event} e
   * @returns {undefined}
   */
  openFilePicker = e => {
    openFilePicker(e, this.filePickerRef);
  };

  /**
   * @returns {JSX.Element} RecipeCreationForm
   */
  render() {
    const {
      props: { recipeCreation },
      state: {
        cookingTime,
        ingredients,
        preparationTime,
        stepAddHover,
        stepIndex,
        steps,
        title,
        videoList,
      },
    } = this;

    return (
      <form className="recipe-creation-form" onSubmit={this.handleSubmit}>
        <div className="container">
          <div className="recipe-creation-body">
            <h2 className="col-title">Add New Recipe</h2>
            <div className="row recipe-header">
              <div className="input-block">
                <input
                  className="text-field text-field--small no-right-border"
                  name="title"
                  placeholder="Title"
                  onChange={this.syncField}
                  value={title}
                />
                <small className="err-msg">{this.handleError('title')}</small>
              </div>

              <div className="input-block">
                <input
                  type="number"
                  className="text-field text-field--small no-right-border"
                  name="preparationTime"
                  onChange={this.syncField}
                  placeholder="Preparation time in mins"
                  value={preparationTime}
                />
                <small className="err-msg">
                  {this.handleError('preparationTime')}
                </small>
              </div>
              <div className="input-block">
                <input
                  type="number"
                  className="text-field text-field--small no-right-border"
                  name="cookingTime"
                  onChange={this.syncField}
                  placeholder="Cooking time in mins"
                  value={cookingTime}
                />
                <small className="err-msg">
                  {this.handleError('cookingTime')}
                </small>
              </div>
            </div>
            <div className="row">
              <div className="column">
                <IngredientList
                  ingredients={ingredients}
                  syncRecipeIngredents={this.syncRecipeIngredents}
                  errorHandler={() => this.handleError('ingredients')}
                />
              </div>
              <div className="column middle">
                <div className="ingredient-content">
                  <p className="col-title">Steps</p>
                  {steps.map(({ description, images, key }, index) => (
                    <div key={key} className="ingredient-image-container">
                      <RecipeStep
                        stepNumber={index + 1}
                        stepIndex={stepIndex}
                        index={index}
                        addState={stepAddHover}
                        images={images}
                        addNewStep={() => this.addNewStep(index)}
                        syncStepProp={this.syncStepProp}
                        description={description}
                      />
                      <small className="err-msg">
                        {this.handleError('step')}
                      </small>
                    </div>
                  ))}
                </div>
              </div>
            </div>
            <div>
              <div className="video-section">
                <p className="col-title">Videos</p>
                <div className="video-content">
                  {steps.map(({ videos, key }, index) =>
                    videos.length ? (
                      <div key={key} className="video-widget">
                        <p className="step">Step {index + 1}</p>
                        <video
                          width="320"
                          height="240"
                          controls
                          src={videos[0]}
                        >
                          <track kind="captions" />
                        </video>
                      </div>
                    ) : null
                  )}
                  {videoList.length > 0 && (
                    <div className="video-widget">
                      <p className="step">Full video</p>
                      <video
                        className="full-video"
                        width="400"
                        height="240"
                        controls
                        src={videoList[0]}
                      >
                        <track kind="captions" />
                      </video>
                    </div>
                  )}
                </div>
              </div>
            </div>
          </div>

          <button type="submit" className="btn-primary">
            {recipeCreation.pending ? (
              <div className="spinner" />
            ) : (
              'Submit Recipe'
            )}
          </button>
          <button
            type="button"
            className="btn-secondary"
            onClick={this.openFilePicker}
          >
            Add Full Video <i className="fas fa-video" />
            <input type="file" ref={this.filePickerRef} hidden />
          </button>
        </div>
        {recipeCreation.pending && (
          <div className="loading-overlay">
            <div className="spinner" />
          </div>
        )}
      </form>
    );
  }
}

RecipeCreationForm.propTypes = {
  recipeCreation: PropTypes.shape({
    penidng: PropTypes.bool,
    created: PropTypes.bool,
    recipe: PropTypes.shape(),
    errors: PropTypes.shape(),
  }).isRequired,
  handleCreation: PropTypes.func.isRequired,
};

export default RecipeCreationForm;
