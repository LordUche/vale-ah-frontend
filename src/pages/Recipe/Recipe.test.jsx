import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import Recipe from './Recipe';

configure({ adapter: new Adapter() });
describe('<Recipe />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<Recipe />);
    expect(toJson(tree)).toMatchSnapshot();
  });
});
