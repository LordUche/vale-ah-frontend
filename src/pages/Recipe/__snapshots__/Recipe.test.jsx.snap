// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe /> matches the snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
